src/shared/find-esp.c

```diff
@@ -497,37 +482,65 @@ int find_esp_and_warn(
                 goto found;
         }
 
-        FOREACH_STRING(dir, "/efi", "/boot", "/boot/efi") {
-                r = chase(dir, root, CHASE_PREFIX_ROOT, &p, NULL);
-                if (r == -ENOENT)
-                        continue;
-                if (r < 0)
-                        return log_error_errno(r,
-                                               "Failed to resolve path %s%s%s: %m",
-                                               dir,
-                                               root ? " under directory " : "",
-                                               strempty(root));
+        FOREACH_STRING(dir, "efi", "boot", "boot/efi") {
+                path = dir;
 
-                r = verify_esp(p, ret_part, ret_pstart, ret_psize, ret_uuid, ret_devid,
+                r = verify_esp(rfd, path, ret_part, ret_pstart, ret_psize, ret_uuid, ret_devid,
                                flags | VERIFY_ESP_SEARCHING);
                 if (r >= 0)
                         goto found;
                 if (!IN_SET(r, -ENOENT, -EADDRNOTAVAIL, -ENOTDIR, -ENOTTY)) /* This one is not it */
                         return r;
-
-                p = mfree(p);
         }
 
         /* No logging here */
         return -ENOKEY;
 
 found:
-        if (ret_path)
-                *ret_path = TAKE_PTR(p);
+        if (ret_path) {
+                r = chaseat(rfd, path, CHASE_AT_RESOLVE_IN_ROOT, ret_path, NULL);
```

Comment:
Please do not move this before `verify_esp()`. I.e. always call this earlier, and drop chase() variants in verify_esp() or friends. Otherwise,
- when this failed, `ret_xyz` partially set.
- the path may be chased multiple times. That sounds redundant and inefficient.